[UnifedPDF] Factor incremental PDF loading code into its own class

smfr opened the pull request:
#### b6ac61e0064cad757dd1a2a81265150c436fe00c
<pre>
[UnifedPDF] Factor incremental PDF loading code into its own class
<a href="https://bugs.webkit.org/show_bug.cgi?id=267979">https://bugs.webkit.org/show_bug.cgi?id=267979</a>
<a href="https://rdar.apple.com/121487714">rdar://121487714</a>

Reviewed by Tim Horton.

Move the incremental PDF loading code out of PDFPlugin into its own class, PDFIncrementalLoader.
All the complexities of ByteRangeRequest and PDFPluginStreamLoaderClient are now internal to that class.
PDFPlugin makes an instance of this class when incremental loading is enabled.

In addition to moving the code, a lot of code cleanup was done:
* Use of thread-safe weak pointers
* More main thread assertions
* Type for ByteRangeRequest identifiers
* Encapsulate the CFDataRef inside PDFPluginBase
* Clearer distinction between data loaded via the main resource request, and byte-range requests

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h: Added.
(WebKit::PDFIncrementalLoader::incrementThreadsWaitingOnCallback):
(WebKit::PDFIncrementalLoader::decrementThreadsWaitingOnCallback):
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm: Added.
(WebKit::ByteRangeRequest::ByteRangeRequest):
(WebKit::ByteRangeRequest::streamLoader):
(WebKit::ByteRangeRequest::setStreamLoader):
(WebKit::ByteRangeRequest::addData):
(WebKit::ByteRangeRequest::position const):
(WebKit::ByteRangeRequest::count const):
(WebKit::ByteRangeRequest::accumulatedData const):
(WebKit::ByteRangeRequest::clearStreamLoader):
(WebKit::ByteRangeRequest::completeWithBytes):
(WebKit::ByteRangeRequest::completeWithAccumulatedData):
(WebKit::ByteRangeRequest::completeIfPossible):
(WebKit::ByteRangeRequest::completeUnconditionally):
(WebKit::PDFPluginStreamLoaderClient::PDFPluginStreamLoaderClient):
(WebKit::PDFPluginStreamLoaderClient::willSendRequest):
(WebKit::PDFPluginStreamLoaderClient::didReceiveResponse):
(WebKit::PDFPluginStreamLoaderClient::didReceiveData):
(WebKit::PDFPluginStreamLoaderClient::didFail):
(WebKit::PDFPluginStreamLoaderClient::didFinishLoading):
(WebKit::PDFIncrementalLoader::create):
(WebKit::PDFIncrementalLoader::PDFIncrementalLoader):
(WebKit::PDFIncrementalLoader::clear):
(WebKit::PDFIncrementalLoader::receivedNonLinearizedPDFSentinel):
(WebKit::PDFIncrementalLoader::documentFinishedLoading const):
(WebKit::PDFIncrementalLoader::ensureDataBufferLength):
(WebKit::PDFIncrementalLoader::appendAccumulatedDataToDataBuffer):
(WebKit::PDFIncrementalLoader::availableDataSize const):
(WebKit::PDFIncrementalLoader::dataPtrForRange const):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidFinishLoading):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidReceiveData):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidFail):
(WebKit::PDFIncrementalLoader::unconditionalCompleteOutstandingRangeRequests):
(WebKit::PDFIncrementalLoader::getResourceBytesAtPositionAfterLoadingComplete):
(WebKit::PDFIncrementalLoader::getResourceBytesAtPosition):
(WebKit::PDFIncrementalLoader::streamLoaderDidStart):
(WebKit::PDFIncrementalLoader::byteRangeRequestForStreamLoader):
(WebKit::PDFIncrementalLoader::forgetStreamLoader):
(WebKit::PDFIncrementalLoader::cancelAndForgetStreamLoader):
(WebKit::PDFIncrementalLoader::identifierForLoader):
(WebKit::PDFIncrementalLoader::removeOutstandingByteRangeRequest):
(WebKit::PDFIncrementalLoader::requestCompleteIfPossible):
(WebKit::PDFIncrementalLoader::requestDidCompleteWithBytes):
(WebKit::PDFIncrementalLoader::requestDidCompleteWithAccumulatedData):
(WebKit::dataProviderGetByteRangesCallback):
(WebKit::dataProviderGetBytesAtPositionCallback):
(WebKit::dataProviderReleaseInfoCallback):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
(WebKit::PDFIncrementalLoader::dataProviderGetByteRanges):
(WebKit::PDFIncrementalLoader::transitionToMainThreadDocument):
(WebKit::PDFIncrementalLoader::threadEntry):
(WebKit::PDFIncrementalLoader::pdfLog):
(WebKit::PDFIncrementalLoader::logStreamLoader):
(WebKit::PDFIncrementalLoader::logState):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):
(WebKit::PDFPlugin::teardown):
(WebKit::PDFPlugin::pdfLog): Deleted.
(WebKit::PDFPlugin::logStreamLoader): Deleted.
(WebKit::PDFPlugin::verboseLog): Deleted.
(WebKit::PDFPlugin::receivedNonLinearizedPDFSentinel): Deleted.
(WebKit::dataProviderGetBytesAtPositionCallback): Deleted.
(WebKit::dataProviderGetByteRangesCallback): Deleted.
(WebKit::dataProviderReleaseInfoCallback): Deleted.
(WebKit::PDFPlugin::maybeClearHighLatencyDataProviderFlag): Deleted.
(WebKit::PDFPlugin::threadEntry): Deleted.
(WebKit::PDFPlugin::unconditionalCompleteOutstandingRangeRequests): Deleted.
(WebKit::PDFPlugin::getResourceBytesAtPositionMainThread): Deleted.
(WebKit::PDFPlugin::getResourceBytesAtPosition): Deleted.
(WebKit::PDFPlugin::adoptBackgroundThreadDocument): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::clearStreamLoader): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeWithBytes): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeWithAccumulatedData): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::maybeComplete): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeUnconditionally): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::willSendRequest): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didReceiveResponse): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didReceiveData): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didFail): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didFinishLoading): Deleted.
(WebKit::PDFPlugin::byteRangeRequestForLoader): Deleted.
(WebKit::PDFPlugin::forgetLoader): Deleted.
(WebKit::PDFPlugin::cancelAndForgetLoader): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidFinishLoading): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidReceiveData): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidFail): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::incrementalPDFLoadingEnabled const):
(WebKit::PDFPluginBase::documentFinishedLoading const):
(WebKit::PDFPluginBase::streamedBytes const):
(WebKit::PDFPluginBase::incrementalPDFStreamDidReceiveData): Deleted.
(WebKit::PDFPluginBase::incrementalPDFStreamDidFinishLoading): Deleted.
(WebKit::PDFPluginBase::incrementalPDFStreamDidFail): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::teardown):
(WebKit::PDFPluginBase::haveStreamedDataForRange const):
(WebKit::PDFPluginBase::haveDataForRange const):
(WebKit::PDFPluginBase::copyDataAtPosition const):
(WebKit::PDFPluginBase::dataPtrForRange const):
(WebKit::PDFPluginBase::insertRangeRequestData):
(WebKit::PDFPluginBase::streamDidReceiveData):
(WebKit::PDFPluginBase::streamDidFinishLoading):
(WebKit::PDFPluginBase::streamDidFail):
(WebKit::PDFPluginBase::adoptBackgroundThreadDocument):
(WebKit::PDFPluginBase::maybeClearHighLatencyDataProviderFlag):
(WebKit::PDFPluginBase::startByteRangeRequest):
(WebKit::PDFPluginBase::receivedNonLinearizedPDFSentinel):
(WebKit::PDFPluginBase::pdfLog):
(WebKit::PDFPluginBase::verboseLog):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: <a href="https://commits.webkit.org/273506@main">https://commits.webkit.org/273506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b925e66f1ee16b782f15e9826fa5c0ef3771d3e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35575 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32055 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11544 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30865 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36128 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12262 "Found 1 new test failure: fast/forms/ios/autocapitalize-words.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31665 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10773 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10789 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39561 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32325 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32138 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36744 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10972 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34804 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12685 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8134 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11482 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h:
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
gotta persist the old years

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h:
```diff
@@ -0,0 +1,137 @@
+    std::atomic<size_t> m_completedRangeRequests { 0 };
+    std::atomic<size_t> m_completedNetworkRangeRequests { 0 };
+#endif
+
```
extra newlines

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
```diff
@@ -230,90 +213,6 @@ class PDFPlugin final : public PDFPluginBase {
 
     URL m_sourceURL;
-
-#if HAVE(INCREMENTAL_PDF_APIS)
```
This makes me so happy

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -195,7 +258,12 @@
 
-    incrementalPDFStreamDidReceiveData(buffer);
+#if !LOG_DISABLED
+    pdfLog(makeString("PDFPluginBase::streamDidReceiveData() - received ", buffer.size(), " bytes, total streamed bytes ", m_streamedBytes));
```
What's with the bespoke logging thing. Is this normal? And the fact that it depends on makeString instead of itself being variadic, it's a bit unusual

smfr replied:
Yes

smfr committed:
[UnifedPDF] Factor incremental PDF loading code into its own class
https://bugs.webkit.org/show_bug.cgi?id=267979
rdar://121487714

Reviewed by Tim Horton.

Move the incremental PDF loading code out of PDFPlugin into its own class, PDFIncrementalLoader.
All the complexities of ByteRangeRequest and PDFPluginStreamLoaderClient are now internal to that class.
PDFPlugin makes an instance of this class when incremental loading is enabled.

In addition to moving the code, a lot of code cleanup was done:
* Use of thread-safe weak pointers
* More main thread assertions
* Type for ByteRangeRequest identifiers
* Encapsulate the CFDataRef inside PDFPluginBase
* Clearer distinction between data loaded via the main resource request, and byte-range requests

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.h: Added.
(WebKit::PDFIncrementalLoader::incrementThreadsWaitingOnCallback):
(WebKit::PDFIncrementalLoader::decrementThreadsWaitingOnCallback):
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm: Added.
(WebKit::ByteRangeRequest::ByteRangeRequest):
(WebKit::ByteRangeRequest::streamLoader):
(WebKit::ByteRangeRequest::setStreamLoader):
(WebKit::ByteRangeRequest::addData):
(WebKit::ByteRangeRequest::position const):
(WebKit::ByteRangeRequest::count const):
(WebKit::ByteRangeRequest::accumulatedData const):
(WebKit::ByteRangeRequest::clearStreamLoader):
(WebKit::ByteRangeRequest::completeWithBytes):
(WebKit::ByteRangeRequest::completeWithAccumulatedData):
(WebKit::ByteRangeRequest::completeIfPossible):
(WebKit::ByteRangeRequest::completeUnconditionally):
(WebKit::PDFPluginStreamLoaderClient::PDFPluginStreamLoaderClient):
(WebKit::PDFPluginStreamLoaderClient::willSendRequest):
(WebKit::PDFPluginStreamLoaderClient::didReceiveResponse):
(WebKit::PDFPluginStreamLoaderClient::didReceiveData):
(WebKit::PDFPluginStreamLoaderClient::didFail):
(WebKit::PDFPluginStreamLoaderClient::didFinishLoading):
(WebKit::PDFIncrementalLoader::create):
(WebKit::PDFIncrementalLoader::PDFIncrementalLoader):
(WebKit::PDFIncrementalLoader::clear):
(WebKit::PDFIncrementalLoader::receivedNonLinearizedPDFSentinel):
(WebKit::PDFIncrementalLoader::documentFinishedLoading const):
(WebKit::PDFIncrementalLoader::ensureDataBufferLength):
(WebKit::PDFIncrementalLoader::appendAccumulatedDataToDataBuffer):
(WebKit::PDFIncrementalLoader::availableDataSize const):
(WebKit::PDFIncrementalLoader::dataPtrForRange const):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidFinishLoading):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidReceiveData):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidFail):
(WebKit::PDFIncrementalLoader::unconditionalCompleteOutstandingRangeRequests):
(WebKit::PDFIncrementalLoader::getResourceBytesAtPositionAfterLoadingComplete):
(WebKit::PDFIncrementalLoader::getResourceBytesAtPosition):
(WebKit::PDFIncrementalLoader::streamLoaderDidStart):
(WebKit::PDFIncrementalLoader::byteRangeRequestForStreamLoader):
(WebKit::PDFIncrementalLoader::forgetStreamLoader):
(WebKit::PDFIncrementalLoader::cancelAndForgetStreamLoader):
(WebKit::PDFIncrementalLoader::identifierForLoader):
(WebKit::PDFIncrementalLoader::removeOutstandingByteRangeRequest):
(WebKit::PDFIncrementalLoader::requestCompleteIfPossible):
(WebKit::PDFIncrementalLoader::requestDidCompleteWithBytes):
(WebKit::PDFIncrementalLoader::requestDidCompleteWithAccumulatedData):
(WebKit::dataProviderGetByteRangesCallback):
(WebKit::dataProviderGetBytesAtPositionCallback):
(WebKit::dataProviderReleaseInfoCallback):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
(WebKit::PDFIncrementalLoader::dataProviderGetByteRanges):
(WebKit::PDFIncrementalLoader::transitionToMainThreadDocument):
(WebKit::PDFIncrementalLoader::threadEntry):
(WebKit::PDFIncrementalLoader::pdfLog):
(WebKit::PDFIncrementalLoader::logStreamLoader):
(WebKit::PDFIncrementalLoader::logState):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::PDFPlugin):
(WebKit::PDFPlugin::teardown):
(WebKit::PDFPlugin::pdfLog): Deleted.
(WebKit::PDFPlugin::logStreamLoader): Deleted.
(WebKit::PDFPlugin::verboseLog): Deleted.
(WebKit::PDFPlugin::receivedNonLinearizedPDFSentinel): Deleted.
(WebKit::dataProviderGetBytesAtPositionCallback): Deleted.
(WebKit::dataProviderGetByteRangesCallback): Deleted.
(WebKit::dataProviderReleaseInfoCallback): Deleted.
(WebKit::PDFPlugin::maybeClearHighLatencyDataProviderFlag): Deleted.
(WebKit::PDFPlugin::threadEntry): Deleted.
(WebKit::PDFPlugin::unconditionalCompleteOutstandingRangeRequests): Deleted.
(WebKit::PDFPlugin::getResourceBytesAtPositionMainThread): Deleted.
(WebKit::PDFPlugin::getResourceBytesAtPosition): Deleted.
(WebKit::PDFPlugin::adoptBackgroundThreadDocument): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::clearStreamLoader): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeWithBytes): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeWithAccumulatedData): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::maybeComplete): Deleted.
(WebKit::PDFPlugin::ByteRangeRequest::completeUnconditionally): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::willSendRequest): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didReceiveResponse): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didReceiveData): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didFail): Deleted.
(WebKit::PDFPlugin::PDFPluginStreamLoaderClient::didFinishLoading): Deleted.
(WebKit::PDFPlugin::byteRangeRequestForLoader): Deleted.
(WebKit::PDFPlugin::forgetLoader): Deleted.
(WebKit::PDFPlugin::cancelAndForgetLoader): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidFinishLoading): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidReceiveData): Deleted.
(WebKit::PDFPlugin::incrementalPDFStreamDidFail): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::incrementalPDFLoadingEnabled const):
(WebKit::PDFPluginBase::documentFinishedLoading const):
(WebKit::PDFPluginBase::streamedBytes const):
(WebKit::PDFPluginBase::incrementalPDFStreamDidReceiveData): Deleted.
(WebKit::PDFPluginBase::incrementalPDFStreamDidFinishLoading): Deleted.
(WebKit::PDFPluginBase::incrementalPDFStreamDidFail): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
(WebKit::PDFPluginBase::teardown):
(WebKit::PDFPluginBase::haveStreamedDataForRange const):
(WebKit::PDFPluginBase::haveDataForRange const):
(WebKit::PDFPluginBase::copyDataAtPosition const):
(WebKit::PDFPluginBase::dataPtrForRange const):
(WebKit::PDFPluginBase::insertRangeRequestData):
(WebKit::PDFPluginBase::streamDidReceiveData):
(WebKit::PDFPluginBase::streamDidFinishLoading):
(WebKit::PDFPluginBase::streamDidFail):
(WebKit::PDFPluginBase::adoptBackgroundThreadDocument):
(WebKit::PDFPluginBase::maybeClearHighLatencyDataProviderFlag):
(WebKit::PDFPluginBase::startByteRangeRequest):
(WebKit::PDFPluginBase::receivedNonLinearizedPDFSentinel):
(WebKit::PDFPluginBase::pdfLog):
(WebKit::PDFPluginBase::verboseLog):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: https://commits.webkit.org/273506@main
